### PR TITLE
[SYCLomatic][DPCT] io_iterator_pair and associated sort APIs

### PR DESCRIPTION
**Pull request by danhoeflinger**
This add the struct io_iterator_pair to the dpct header, and and adds associated APIs for sort_keys, sort_pairs, and segmented_sort_pairs.  

Signed-off-by: Dan Hoeflinger <dan.hoeflinger@intel.com>

### CODE REVIEW

File: clang/runtime/dpct-rt/include/dpl_extras/algorithm.h.inc
Comment (jzc) [inline clang/runtime/dpct-rt/include/dpl_extras/algorithm.h.inc]: `bool descending = false` to match the other overload of `sort_keys`?

File: clang/runtime/dpct-rt/include/dpl_extras/iterators.h.inc
Comment (jzc) [inline clang/runtime/dpct-rt/include/dpl_extras/iterators.h.inc]: `__selector` and `__iter` are both public, is this intended? I would not expect public members to be prefixed by double underscores.

**Review by jzc — CHANGES_REQUESTED** [review]
The following for me has a compile error:

```
void sort(int *k_in, int *k_out, int n) {
  dpct::io_iterator_pair keys(k_in, k_out);
  dpct::sort_keys(oneapi::dpl::execution::device_policy(dpct::get_default_queue()),
		  keys, n, false);
}
```

```
error: call to 'sort_keys' is ambiguous
  dpct::sort_keys(oneapi::dpl::execution::device_policy(dpct::get_default_queue()),

algorithm.h:1376:13: note: candidate function [with _ExecutionPolicy = oneapi::dpl::execution::device_policy<>, key_t = dpct::io_iterator_pair<int *>, key_out_t = int]

algorithm.h:1409:13: note: candidate function [with _ExecutionPolicy = oneapi::dpl::execution::device_policy<>, key_t = int *]
inline void sort_keys(
```

It seems the other overload for `sort_keys` needs to be disabled when the type of its 2nd/3rd argument is not an iterator.

File: clang/runtime/dpct-rt/include/dpl_extras/algorithm.h.inc
Comment (danhoeflinger) [reply clang/runtime/dpct-rt/include/dpl_extras/algorithm.h.inc]: Good catch, fixed.

File: clang/runtime/dpct-rt/include/dpl_extras/iterators.h.inc
Comment (danhoeflinger) [reply clang/runtime/dpct-rt/include/dpl_extras/iterators.h.inc]: Fixed naming convention, these are intended to be public.

**Comment by danhoeflinger** [comment]
>It seems the other overload for sort_keys needs to be disabled when the type of its 2nd/3rd argument is not an iterator.

I resolved this, and modified the test in https://github.com/oneapi-src/SYCLomatic-test/pull/139 to expose this issue as well.  Thanks for catching this.

**Review by jzc — APPROVED** [review]
LGTM

File: clang/runtime/dpct-rt/include/dpl_extras/algorithm.h.inc
Comment (LU-JOHN) [inline clang/runtime/dpct-rt/include/dpl_extras/algorithm.h.inc:52]: How are vectors being used in this commit?  I couldn't find a use.

Comment (danhoeflinger) [reply clang/runtime/dpct-rt/include/dpl_extras/algorithm.h.inc:52]: The helper `dpct::internal::is_iterator<T>` is used to help avoid ambiguous function signatures.

**Review by LU-JOHN — APPROVED** [review]

File: clang/runtime/dpct-rt/include/dpl_extras/algorithm.h.inc
Comment (danhoeflinger) [reply clang/runtime/dpct-rt/include/dpl_extras/algorithm.h.inc:52]: Wait before merging... I just realized I forgot to include in the .inc file the `is_iterator` dependency.

Comment (danhoeflinger) [reply clang/runtime/dpct-rt/include/dpl_extras/algorithm.h.inc:52]: I've added the missing dependencies now, sorry about that.

**Comment by danhoeflinger** [comment]
I believe my changes to the dependencies in the .inc file caused some changes in the api_tests.  However, resolving this is beyond the scope of my expertise.  Can someone help adjust these tests? @yihanwg perhaps?

**Comment by jzc** [comment]
@danhoeflinger The api_test* migrate some file, and find the number of lines in the generated `MainSourceFiles.yaml` with the string `IsCalled`. My current understanding of this is that a line with `IsCalled` is a necessary feature of the header files the migration requires, so the number of lines found represents the total number of necessary features. Since you added a dependency to the `sort_keys` methods (and that feature itself has no other dependent features), a migration requesting `sort_keys` will need one more feature (namely, the `is_iterator` feature) than it did before. So you can update the `// CHECK: n` in the failing tests to `// CHECK: n+1`.

**Comment by danhoeflinger** [comment]
> @danhoeflinger The api_test* migrate some file, and find the number of lines in the generated `MainSourceFiles.yaml` with the string `IsCalled`. My current understanding of this is that a line with `IsCalled` is a necessary feature of the header files the migration requires, so the number of lines found represents the total number of necessary features. Since you added a dependency to the `sort_keys` methods (and that feature itself has no other dependent features), a migration requesting `sort_keys` will need one more feature (namely, the `is_iterator` feature) than it did before. So you can update the `// CHECK: n` in the failing tests to `// CHECK: n+1`.

That makes sense.  I'll look into resolving this in the afternoon.  Thanks.

**Comment by danhoeflinger** [comment]
The api_tests should be resolved.  The failing CI seems unrelated, and I do not have permissions to rerun.

**Comment by yihwang-nv** [comment]
> The api_tests should be resolved. The failing CI seems unrelated, and I do not have permissions to rerun.

Please just ignore `module-kernel` and `kernel_library` in Jekins/Windows_CI, It's a known issue, and I have rerun other failed CI.

**Review by yihwang-nv — APPROVED** [review]

**Comment by danhoeflinger** [comment]
Thanks for the approvals.
Just in case you are waiting on me to merge, I do not have permissions to do so.

**Pull request merged**
